app/frontend/src/features/communities/EditCommunityPage.tsx

```diff
@@ -0,0 +1,145 @@
+import { Typography } from "@material-ui/core";
+import Button from "components/Button";
+import MarkdownInput from "components/MarkdownInput";
+import PageTitle from "components/PageTitle";
+import Snackbar from "components/Snackbar";
+import { UPDATE } from "features/constants";
+import { Error as GrpcError } from "grpc-web";
+import { Community } from "pb/communities_pb";
+import { Page } from "pb/pages_pb";
+import { communityKey } from "queryKeys";
+import { useForm } from "react-hook-form";
+import { useMutation, useQueryClient } from "react-query";
+import { Redirect, useHistory } from "react-router-dom";
+import { routeToCommunity } from "routes";
+import { service } from "service";
+import makeStyles from "utils/makeStyles";
+
+import CommunityBase from "./CommunityBase";
+import {
+  COMMUNITY_PAGE_UPDATED,
+  EDIT_COMMUNITY_PAGE,
+  PAGE_CONTENT_FIELD_LABEL,
+} from "./constants";
+
+const useStyles = makeStyles((theme) => ({
+  root: {
+    display: "flex",
+    justifyContent: "space-between",
+  },
+  form: {
+    display: "grid",
+    paddingBottom: theme.spacing(5),
+    rowGap: theme.spacing(1),
+    width: "100%",
+  },
+  updateButton: {
+    justifySelf: "end",
+  },
+}));
+
+interface UpdatePageData {
+  communityId: string;
+  content: string;
+  pageId: string;
+  slug: string;
+}
+
+export default function EditCommunityPage() {
+  const classes = useStyles();
+  const queryClient = useQueryClient();
+  const { control, handleSubmit, register } = useForm<UpdatePageData>();
+  const history = useHistory();
+
+  const {
+    error,
+    isLoading,
+    isSuccess,
+    mutate: updatePage,
+  } = useMutation<Page.AsObject, GrpcError, UpdatePageData>(
+    ({ content, pageId }) =>
+      service.pages.updatePage({ content, pageId: +pageId }),
+    {
+      onSuccess(newPageData, { communityId, slug }) {
+        queryClient.setQueryData<Community.AsObject | undefined>(
```

Comment:
Optimistic update for some reason doesn't work here and I'm not quite sure what I've done wrong compared to other similar queries... on slow network, even the `history.push` redirect you can see the content is updating slowly...

Reply:
Good catch! It looks like `onBlur` and `onChange` from `useController` doesn't have a stable identity and gets re-created on every re-render, and we can't really memoise with `useCallback` since the exhaustive dep rule will just ask `fieldonBlur`/`fieldonChange` to be a dependency which will lead to the same problem as in that `useEffect`...

The other workaround I came up with is storing `fieldOnBlur`/`fieldOnChange` in refs and call that in the `useEffect` that creates the ToastEditor instead. Since neither `fieldOnBlur` nor `fieldOnChange` depends on state, it's probably ok?